media-source-respec.html

```diff
@@ -1128,17 +1222,28 @@ <h2><dfn>SourceBuffer</dfn> Object</h2>
 <pre class="idl">enum AppendMode {
     "segments",
     "sequence"
-};</pre><table class="simple" data-dfn-for="AppendMode"><tbody><tr><th colspan="2">Enumeration description</th></tr><tr><td><dfn><code id="idl-def-AppendMode.segments">segments</code></dfn></td><td>
-          <p>The timestamps in the media segment determine where the [=coded frames=] are placed in the presentation. Media segments can be appended in any order.</p>
-        </td></tr><tr><td><dfn><code id="idl-def-AppendMode.sequence">sequence</code></dfn></td><td>
-          <p>Media segments will be treated as adjacent in time independent of the timestamps in the media segment. Coded frames in a new media segment will be placed immediately after the coded
-            frames in the previous media segment. The {{SourceBuffer/timestampOffset}} attribute will be updated if a new offset is needed to make the new media segments adjacent to the previous media segment.
-            Setting the {{SourceBuffer/timestampOffset}} attribute in {{AppendMode/""sequence""}} mode allows a media segment to be placed at a specific position in the timeline without any knowledge
-            of the timestamps in the media segment.
-          </p>
+};</pre><table class="simple" data-dfn-for="AppendMode"><tbody><tr><th colspan="2">Enumeration description</th></tr>
+      <tr><td><dfn><code id="idl-def-AppendMode.segments">segments</code></dfn></td><td>
+        <p>The timestamps in the media segment or {{EncodedChunks}} determine where the [=coded frames=] are placed in
+        the presentation. Media segments can be appended in any order. Discontinuous {{EncodedChunks}} should not be
+        appended in the same {{SourceBuffer/appendEncodedChunks()}} call: the application should instead wait for the
+        {{SourceBuffer}} to no longer be {{SourceBuffer/updating}}, and then call {{SourceBuffer/abort()}} or
+        {{SourceBuffer/changeType()}} before appending {{EncodedChunks}} that are discontinuous with those in the
+        previous {{SourceBuffer/appendEncodedChunks()}} call.</p>
```

Comment:
I'm curious why EncodedChunks cannot be appended in any order like media segments can ?

Does the above mean that to append, say, overlapping media (replacing some existing frames) we need to call `changeType` even if the type has not changed ?

Reply:
I agree this is not ergonomic/compatible with previous behavior, and should be improved.

If I recall from when I wrote this draft and the prototype in Chromium, sequences of out-of-order presentation time versus the decode sequence must still be supported. The issue, I think, arises from the lack of explicit "decode timestamp" in EncodedChunks, and the prototype/draft API instead assumes the order of chunk appends describes their decode sequence. There is no ability of the coded frame processing algorithm to detect discontinuities in EncodedChunk decode time, unlike bytestream-formatted media. Without some ability to determine actual discontinuity implicitly, explicit discontinuity signaling is needed (abort() or changeType()). Even so, `changeType` isn't required if type isn't changing, but there is indeed a discontinuity. Instead, abort() signals the discontinuity explicitly without changing type.

This could be improved/simplified for "segments" mode: keyframes themselves could be used as the implicit signal of potential discontinuity, especially since there is allowance for only 1 audio config or only 1 video config in a `SourceBuffer` configured to allow EncodedChunk appends (and therefore there is no need to have such per-keyframe discontinuity also impacting the processing of coded frames for any other track buffers in that `SourceBuffer`.)

I might be missing some other detail in my recollection of this old PR and prototype though. For example, "sequence" mode appends that are not discontinuous at keyframes, but might appear to be if they are from SAP-Type-2 encode sequences, could cause interop issues if keyframes are used as implicit potential discontinuities. Conversely, without implicit discontinuity signaling, lack of app explicitly signaling could impair reaching the intent of "sequence" mode buffering.

One solution might be to use keyframes as implicit potential discontinuity signaling, but disallow use of appending chunks in "sequence" mode.